pipe/pipe.sh

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env sh
+#
+set -e
+
+source "$(dirname "$0")/common.sh"
+
+validate() {
+     DEBUG=${DEBUG:=false}
+     SKIP_DEPENDENCIES=${SKIP_DEPENDENCIES:=false}
+     LEVEL=${LEVEL:=0}
+     AUTOLOADER=${AUTOLOADER:="vendor/autoload.php"}
+     IGNORE_PLATFORM_DEPENDENCIES=${IGNORE_PLATFORM_DEPENDENCIES:=false}
+}
+
+setup_ssh_creds() {
+     # Setup pipeline SSH 
+     INJECTED_SSH_CONFIG_DIR="/opt/atlassian/pipelines/agent/ssh"
+     IDENTITY_FILE="${INJECTED_SSH_CONFIG_DIR}/id_rsa_tmp"
+     KNOWN_SERVERS_FILE="${INJECTED_SSH_CONFIG_DIR}/known_hosts"
+     if [ ! -f ${IDENTITY_FILE} ]; then
+          info "No default SSH key configured in Pipelines.\n These are required to install internal composer packages. \n These should be generated in bitbucket settings at Pipelines > SSH Keys."
+          return
+     fi
+     mkdir -p ~/.ssh
+     touch ~/.ssh/authorized_keys
+     cp ${IDENTITY_FILE} ~/.ssh/pipelines_id
+
+     if [ ! -f ${KNOWN_SERVERS_FILE} ]; then
+          fail "No SSH known_hosts configured in Pipelines."
+     fi
+     cat ${KNOWN_SERVERS_FILE} >> ~/.ssh/known_hosts
+     if [ -f ~/.ssh/config ]; then
+          debug "Appending to existing ~/.ssh/config file"
+     fi
+     echo "IdentityFile ~/.ssh/pipelines_id" >> ~/.ssh/config
+     chmod -R go-rwx ~/.ssh/
+}
+
+inject_composer_creds() {
+     if [[ -z "${MAGENTO_USER}" ]] | [[ -z "${MAGENTO_PASS}" ]]; then
+          info "No Magento Composer details configured. Skiping."
+     else
+          echo "Injecting Magento Composer credentials into auth.json"
+          jq '."http-basic"."repo.magento.com".username = env.MAGENTO_USER | ."http-basic"."repo.magento.com".password = env.MAGENTO_PASS | del(."github-oauth")' auth.json.sample > auth.json
+          if $DEBUG; then
+               echo "auth.json"
+               cat auth.json
+          fi
+     fi
+}
+
+
+install_composer_dependencies() {
+     echo "Installing composer dependencies"
+     composer install --dev $([ $IGNORE_PLATFORM_DEPENDENCIES ] && printf '--ignore-platform-reqs')
+}
+
+run_phpstan() {
+     debug "Testing modified files in this branch..."
+
+     TARGET_BRANCH='origin/master'
+     if [ -n "$BITBUCKET_PR_DESTINATION_BRANCH" ]; then
+          TARGET_BRANCH="origin/$BITBUCKET_PR_DESTINATION_BRANCH"
+     fi
+
+     if $DEBUG; then
+          echo "State of working directory"
+          git status
+     fi
+
+     echo "Comparing HEAD against branch $TARGET_BRANCH"
+     MERGE_BASE=$(git merge-base HEAD $TARGET_BRANCH)
+
+     CHANGED_FILES=$(git diff --relative --name-only --diff-filter=AM $MERGE_BASE -- '*.php' '*.phtml')
```

Comment:
This process seems to assume PHPStan has been run at least once on the whole repo.

Should that be documented or add a flag to force a full scan?

Reply:
A flag for a full scan is probably a good idea, the idea behind the changed files though is that we don't want the devs to have to go and make a HUGE amount of changes when introducing phpstan. 

Instead if we only do it on new PRs they will slowly get used to it and touch more and more code, then we hit them with the big scan :D